### PR TITLE
Mixpanel

### PR DESCRIPTION
**Pull request by leogrim**

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/UserController.scala [hunk 1]
```diff
@@ -206,10 +209,9 @@ class UserController @Inject() (
...
           val cleanFirst = User.sanitizeName(userData.firstName getOrElse user.firstName)
           val cleanLast = User.sanitizeName(userData.lastName getOrElse user.lastName)
-          userRepo.save(user.copy(
-            firstName = cleanFirst,
-            lastName = cleanLast
-          ))
+          val updatedUser = user.copy(firstName = cleanFirst, lastName = cleanLast)
+          userRepo.save(updatedUser)
+          SafeFuture { heimdal.engageUser(updatedUser) }
         }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/UserController.scala:215]: @andrewconner 
Updating user info in Mixpanel here

File: kifi-backend/modules/shoebox/app/com/keepit/social/SecureSocialUserServiceImpl.scala [hunk 2]
```diff
@@ -178,7 +179,8 @@ class SecureSocialUserPluginImpl @Inject() (
         } else {
           val user = userRepo.get(socialUserInfo.userId.get)
           if (user.state == UserStates.INCOMPLETE_SIGNUP && isComplete) {
-            userRepo.save(user.withName(socialUser.firstName, socialUser.lastName).withState(newUserState))
+            val updatedUser = userRepo.save(user.withName(socialUser.firstName, socialUser.lastName).withState(newUserState))
+            SafeFuture { heimdal.engageUser(updatedUser) }
           }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/social/SecureSocialUserServiceImpl.scala:184]: @andrewconner 
and here

**Comment by leogrim** [comment]
@andrewconner 
I have tried to understand when users are created / updated so that we can send that information to Mixpanel, hope the hooks I have added make sense and that I haven't missed anything (email signup?)

**Comment by andrewconner** [comment]
Could you just override the `save` method in `UserRepo`? Not sure we'll remember to do this every time we change a user.

**Comment by leogrim** [comment]
It's the easy way, just thought it was kind of messy to have Mixpanel at the repo level, no?

**Comment by andrewconner** [comment]
That's a good point. We just don't have any checks that ensure we keep this pattern up every time we mutate user, so it might make sense to put it there.

**Comment by leogrim** [comment]
The other reason why I backed off from that is because userRepo is in our all tests, which means adding HeimdalServiceClientModule everywhere. But that's probably safer anyway :)

**Comment by andrewconner** [comment]
Ah. You can have a fake one that tests, but yeah. Should only affect shoebox tests though, fortunately.

**Comment by leogrim** [comment]
Yeah yeah yeah, we already have a fake one, it's just polluting the code
that's all

2013/11/18 Andrew Conner notifications@github.com

**Comment by leogrim** [comment]
Oh not that much actually, thanks to ShoeboxApplication and
ShoeboxTestInjector the dependencies are quite localized

2013/11/18 Léo Grimaldi leo.grimaldi.fr@gmail.com

**Pull request merged**